indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java

```diff
@@ -1021,92 +1043,164 @@ public void unregisterListener(String listenerId)
+      pendingTasksExec.submit(
+          () -> {
             try {
-              synchronized (statusLock) {
-                taskItem = tasks.get(taskId);
+              if (!lifecycleLock.awaitStarted()) {
+                log.makeAlert("Lifecycle not started, PendingTaskExecution loop will not run.").emit();
+                return;
+              }
 
-                if (taskItem == null) {
-                  log.info(
-                      "Task[%s] work item not found. Probably user asked to shutdown before. Not assigning.",
-                      taskId
-                  );
-                  return;
-                }
+              pendingTasksExecutionLoop();
+            }
+            catch (Throwable t) {
+              log.makeAlert(t, "Error while waiting for lifecycle start. PendingTaskExecution loop will not run")
+                 .emit();
+            }
+            finally {
+              log.info("PendingTaskExecution loop exited.");
+            }
+          }
+      );
+    }
+  }
 
-                if (taskItem.getState() != HttpRemoteTaskRunnerWorkItem.State.PENDING) {
-                  log.info(
-                      "Task[%s] is in state[%s]. Probably some worker already reported it. Not assigning.",
-                      taskId,
-                      taskItem.getState()
-                  );
-                  return;
-                }
+  private void pendingTasksExecutionLoop()
+  {
+    while (!Thread.interrupted() && lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS)) {
+      try {
+        // Find one pending task to run and a worker to run on
+        HttpRemoteTaskRunnerWorkItem taskItem = null;
+        ImmutableWorkerInfo immutableWorker = null;
 
-                if (taskItem.getTask() == null) {
-                  throw new ISE("WTF! couldn't find Task instance for taskId[%s].", taskId);
-                }
-                immutableWorker = findWorkerToRunTask(taskItem.getTask());
-
-                if (immutableWorker == null) {
-                  // no free worker, wait for some worker to become free
-                  statusLock.wait(config.getWaitForWorkerSlot().toStandardDuration().getMillis());
-                  continue;
-                } else if (workersWithUnacknowledgedTask.putIfAbsent(
-                    immutableWorker.getWorker().getHost(),
-                    taskId
-                ) != null) {
-                  // there was a race and someone else took this worker slot, try again
-                  continue;
-                }
-              }
+        synchronized (statusLock) {
+          Iterator<String> iter = pendingTaskIds.iterator();
+          while (iter.hasNext()) {
+            String taskId = iter.next();
+            HttpRemoteTaskRunnerWorkItem ti = tasks.get(taskId);
+
+            if (ti == null || !ti.getState().isPending()) {
+              // happens if the task was shutdown or was picked up earlier and no more pending
+              iter.remove();
```

Comment:
Looks like `pendingTaskIds` is updated only here. This will lead to processing the same taskId multiple times. HRTR will assign the task to some worker in the first loop, ignore it until the worker starts processing that task in the below `if` clause, and then finally remove the taskId here. Can we update `pendingTaskIds` when the task is successfully assigned?

Reply:
before leaving the lock , ti state is updated to PENDING_WORKER_ASSIGN so no other task executor thread can pick up.

`pendingTaskIds` variable declaration has a comment that this variable is exclusively manipulated by only external task submitter threads or task executor threads which I preferred.

Reply:
Hmm, sorry I don't think I fully understand this logic yet. 

> before leaving the lock , ti state is updated to PENDING_WORKER_ASSIGN so no other task executor thread can pick up.

Do you mean the other threads can pick up even after the ti state is set to `PENDING_WORKER_ASSIGN` but will ignore it in the [below if clause](https://github.com/apache/incubator-druid/pull/8697/files#diff-b994d2f0b67b07608a060f52a17fbd01R1107)? Or, do they really not pick up because somehow `pendingTaskIds` is updated before other threads start picking up?

> `pendingTaskIds` variable declaration has a comment that this variable is exclusively manipulated by only external task submitter threads or task executor threads which I preferred.

Yes, I think it makes sense and I prefer that too. But can `pendingTaskIds` be updated after [this line](https://github.com/apache/incubator-druid/pull/8697/files#diff-b994d2f0b67b07608a060f52a17fbd01R1141)? Like as the below code:

```java
            // set state to PENDING_WORKER_ASSIGN before releasing the lock so that this task item is not picked
            // up by another task execution thread.
            ti.setState(HttpRemoteTaskRunnerWorkItem.State.PENDING_WORKER_ASSIGN);
            iter.remove(); // new line to remove it from pendingTaskIds immediately
            taskItem = ti;
            break;
```

I don't think this is enough since the task id should be added back to `pendingTaskIds` if the worker couldn't execute the assigned task, but seems possible.

Reply:
> Do you mean the other threads can pick up even after the ti state is set to PENDING_WORKER_ASSIGN but will ignore it in the below if clause? Or, do they really not pick up because somehow pendingTaskIds is updated before other threads start picking up?

yes, other executor threads would see that but ignore because of the if clause you noted.

> But can pendingTaskIds be updated after this line? Like as the below code: ...

if we removed here then we might have to add it back just in case task couldn't run for whatever reason, adding back would be complex because we will find out that we need to add it back at some later time and also adding back to its old position in the list would be difficult.
as opposed to all that we remove it from `pendintTaskIds` only when task is known to be no longer in pending state.

Reply:
if you still have doubts, maybe describe a scenario where you think this doesn't work and I will try to explain how I think that scenario would play out :) .

Reply:
Oh I believe this would work. I was just curious whether it was intended or not. I'm ok with it if this was intended. Would you please add a comment to avoid confusion in the future?

Reply:
got it. sure, added some commentary to hopefully make it clearer.